NYC 311 工单 SQL 分析

当前只使用首个测试分片，覆盖 2024-09-07 至 2024-09-30。

连接 MySQL

In [1]:
%load_ext sql
import sys
from pathlib import Path
sys.path.insert(0, str((Path.cwd().parent / '自动化实现').resolve()))
from 自动化入库 import connect_mysql

engine = connect_mysql()
%sql engine
%config SqlMagic.displaylimit = 50


先看当前数据量和日期范围。

In [2]:
%%sql
SELECT
    COUNT(*) AS work_order_count,
    MIN(created_date) AS first_created_date,
    MAX(created_date) AS last_created_date
FROM fact_work_order;

Running query in 'mysql+pymysql://root:***@localhost:3306/nyc_311_project?charset=utf8mb4'

1 rows affected.

work_order_count,first_created_date,last_created_date
248139,2024-09-07 00:00:00,2024-09-30 23:59:31


**结果：** 当前测试库共有 248,139 条工单，创建时间从 2024-09-07 00:00:00 到 2024-09-30 23:59:31。

1. 不同机构主要处理哪些工单？

In [3]:
%%sql
-- 统计各机构的工单类型和数量。
WITH complaint_counts AS (
    SELECT agency, complaint_type, COUNT(*) AS work_order_count
    FROM fact_work_order
    WHERE complaint_type IS NOT NULL
    GROUP BY agency, complaint_type
),
-- 计算机构内部占比，以及该机构占该类工单的比例。
complaint_shares AS (
    SELECT
        c.agency,
        c.complaint_type,
        c.work_order_count,
        SUM(c.work_order_count) OVER (PARTITION BY c.agency) AS agency_total,
        SUM(c.work_order_count) OVER (PARTITION BY c.complaint_type) AS complaint_total,
        ROUND(c.work_order_count / SUM(c.work_order_count)
            OVER (PARTITION BY c.agency) * 100, 2) AS agency_percentage,
        ROUND(c.work_order_count / SUM(c.work_order_count)
            OVER (PARTITION BY c.complaint_type) * 100, 2) AS complaint_agency_percentage
    FROM complaint_counts AS c
)
-- 连接机构全称，保留全部机构和工单类型明细。
SELECT
    s.agency,
    d.agency_name_standard,
    s.complaint_type,
    s.work_order_count,
    s.agency_total,
    s.complaint_total,
    s.agency_percentage,
    s.complaint_agency_percentage
FROM complaint_shares AS s
LEFT JOIN dim_agency AS d ON s.agency = d.agency
ORDER BY s.work_order_count DESC;

Running query in 'mysql+pymysql://root:***@localhost:3306/nyc_311_project?charset=utf8mb4'

171 rows affected.

agency,agency_name_standard,complaint_type,work_order_count,agency_total,complaint_total,agency_percentage,complaint_agency_percentage
NYPD,New York City Police Department,Noise - Residential,39381,131965,39381,29.84,100.00
NYPD,New York City Police Department,Illegal Parking,37255,131965,37255,28.23,100.00
NYPD,New York City Police Department,Noise - Street/Sidewalk,16795,131965,16795,12.73,100.00
NYPD,New York City Police Department,Blocked Driveway,12317,131965,12317,9.33,100.00
HPD,Department of Housing Preservation and Development,UNSANITARY CONDITION,8838,33705,8838,26.22,100.00
NYPD,New York City Police Department,Noise - Commercial,5755,131965,5755,4.36,100.00
DSNY,Department of Sanitation,Dirty Condition,5430,24018,5430,22.61,100.00
NYPD,New York City Police Department,Abandoned Vehicle,5105,131965,5105,3.87,100.00
DEP,Department of Environmental Protection,Noise,4515,11406,4515,39.58,100.00
HPD,Department of Housing Preservation and Development,PAINT/PLASTER,4210,33705,4210,12.49,100.00


前50行包含12家机构。数量最大的4个组合都属于NYPD，分别是住宅噪声、违停、街道噪声和车道堵塞；HPD的住房卫生问题、DSNY的环境卫生问题和DEP的噪声也排在前列。

DOE和OTI没有进入前50行，只能说明其组合数量较小，不能据此判断它们没有明确职责。

2. 对同一类工单，不同机构的关闭时间有什么差异？

In [4]:
%%sql
-- 找出由多家机构共同处理的工单类型。
WITH comparable_types AS (
    SELECT complaint_type
    FROM fact_work_order
    WHERE complaint_type IS NOT NULL
    GROUP BY complaint_type
    HAVING COUNT(DISTINCT agency) > 1
)
-- 比较同类工单在不同机构的平均关闭时间。
SELECT
    w.complaint_type,
    w.agency,
    d.agency_name_standard,
    COUNT(*) AS closed_order_count,
    ROUND(AVG(w.closure_duration_minutes) / 60, 2) AS average_closure_hours
FROM fact_work_order AS w
INNER JOIN comparable_types AS c ON w.complaint_type = c.complaint_type
LEFT JOIN dim_agency AS d ON w.agency = d.agency
WHERE w.closure_duration_minutes IS NOT NULL
GROUP BY w.complaint_type, w.agency, d.agency_name_standard
ORDER BY w.complaint_type, closed_order_count DESC;

Running query in 'mysql+pymysql://root:***@localhost:3306/nyc_311_project?charset=utf8mb4'

12 rows affected.

complaint_type,agency,agency_name_standard,closed_order_count,average_closure_hours
Asbestos,DEP,Department of Environmental Protection,51,7.74
Asbestos,DOHMH,Department of Health and Mental Hygiene,37,336.06
Elevator,DOB,Department of Buildings,1306,1400.94
ELEVATOR,HPD,Department of Housing Preservation and Development,96,259.53
Encampment,NYPD,New York City Police Department,3088,2.44
Encampment,DHS,Department of Homeless Services,1199,195.95
Graffiti,DSNY,Department of Sanitation,1282,439.75
Graffiti,NYPD,New York City Police Department,105,1.53
Highway Condition,DOT,Department of Transportation,85,202.41
Highway Condition,DSNY,Department of Sanitation,54,279.74


当前分片只有6种工单由多家机构处理，共形成12个可比较的机构—工单类型组合。

同类工单的平均关闭时间差异很大。这里只描述工单从创建到关闭所经过的时间；机构可能承担不同处理环节，也可能采用不同关闭口径，因此不能直接把差异解释为机构效率高低。

3. 各行政区最常报告哪些问题？同类问题在各区的数量和关闭时间有什么差异？

In [5]:
%%sql
-- 汇总各行政区的工单类型、数量和平均关闭时间。
WITH borough_complaints AS (
    SELECT
        borough,
        complaint_type,
        COUNT(*) AS work_order_count,
        COUNT(closed_date) AS closed_order_count,
        ROUND(AVG(closure_duration_minutes) / 60, 2) AS average_closure_hours
    FROM fact_work_order
    WHERE borough != 'Unspecified' AND complaint_type IS NOT NULL
    GROUP BY borough, complaint_type
)
-- 同时计算区内占比和该区占全市同类工单的比例。
SELECT
    b.*,
    ROUND(b.work_order_count / SUM(b.work_order_count)
        OVER (PARTITION BY b.borough) * 100, 2) AS borough_percentage,
    ROUND(b.work_order_count / SUM(b.work_order_count)
        OVER (PARTITION BY b.complaint_type) * 100, 2) AS complaint_borough_percentage
FROM borough_complaints AS b
ORDER BY b.work_order_count DESC;

Running query in 'mysql+pymysql://root:***@localhost:3306/nyc_311_project?charset=utf8mb4'

694 rows affected.

borough,complaint_type,work_order_count,closed_order_count,average_closure_hours,borough_percentage,complaint_borough_percentage
BRONX,Noise - Residential,21038,21038,3.95,36.12,53.42
BROOKLYN,Illegal Parking,14460,14460,2.61,19.87,38.82
QUEENS,Illegal Parking,11048,11048,4.5,19.16,29.66
BROOKLYN,Noise - Residential,7120,7120,1.38,9.78,18.08
QUEENS,Noise - Residential,5798,5798,2.19,10.06,14.72
MANHATTAN,Illegal Parking,5531,5531,1.75,10.82,14.85
MANHATTAN,Noise - Street/Sidewalk,5227,5227,1.3,10.23,31.12
BRONX,Illegal Parking,5138,5138,10.2,8.82,13.79
QUEENS,Blocked Driveway,4942,4942,4.68,8.57,40.12
BROOKLYN,Blocked Driveway,4907,4907,2.81,6.74,39.84


前50行覆盖5个行政区，其中Brooklyn有17行、Queens有13行、Manhattan有11行、Bronx有8行，Staten Island只有1行。

数量最多的是Bronx的住宅噪声，其次是Brooklyn和Queens的违停。同为违停，前50行中的平均关闭时间从Manhattan的1.75小时到Bronx的10.20小时不等。Staten Island在前50行中信息不足，暂不概括其问题结构。

4. 哪些工单最常见？各具体类型的占比是多少？

In [ ]:
%%sql
-- 保留具体工单类型，分组在报表中选择。
WITH complaint_counts AS (
    SELECT
        COALESCE(complaint_type, 'Unknown') AS complaint_group,
        COUNT(*) AS work_order_count
    FROM fact_work_order
    GROUP BY COALESCE(complaint_type, 'Unknown')
)
SELECT
    complaint_group,
    work_order_count,
    ROUND(work_order_count / SUM(work_order_count) OVER () * 100, 2)
        AS percentage
FROM complaint_counts
ORDER BY work_order_count DESC;

已撤回低频类型合并，保留具体类型。此查询尚未重新运行，原归并结果不再作为当前结论。

5. 哪些工单通常关闭得快，哪些容易拖很久？

In [7]:
%%sql
-- 比较各类工单的关闭率和平均关闭时间。
SELECT
    complaint_type,
    COUNT(*) AS work_order_count,
    COUNT(closed_date) AS closed_order_count,
    ROUND(COUNT(closed_date) / COUNT(*) * 100, 2) AS closed_percentage,
    ROUND(AVG(closure_duration_minutes) / 60, 2) AS average_closure_hours
FROM fact_work_order
WHERE complaint_type IS NOT NULL
GROUP BY complaint_type
ORDER BY average_closure_hours DESC;

Running query in 'mysql+pymysql://root:***@localhost:3306/nyc_311_project?charset=utf8mb4'

165 rows affected.

complaint_type,work_order_count,closed_order_count,closed_percentage,average_closure_hours
New Tree Request,1138,1138,100.00,5965.81
DEP Street Condition,2,2,100.00,4871.41
Borough Office,4,4,100.00,4498.66
Building/Use,1440,1440,100.00,4079.77
Lot Condition,181,179,98.90,3734.93
Root/Sewer/Sidewalk Condition,922,865,93.82,3561.67
Wood Pile Remaining,40,31,77.50,3212.93
Tunnel Condition,2,2,100.00,3064.22
Tattooing,11,11,100.00,2860.2
Poison Ivy,6,6,100.00,2775.49


前50行按平均关闭时间从高到低排列。New Tree Request平均5965.81小时，Building/Use平均4079.77小时，Lot Condition平均3734.93小时，属于前50行中样本量较大的长耗时类型。

部分靠前类型只有2至11条工单，平均值不稳定，因此不能只看排序判断难度。当前前50行只支持观察长耗时类型，不用于概括关闭最快的类型。

6. 短工单和长工单有什么区别？

In [8]:
%%sql
-- NTILE(4) 按关闭耗时把工单平均分成四组。
WITH duration_orders AS (
    SELECT
        complaint_type,
        agency,
        borough,
        open_data_channel_type,
        DAYOFWEEK(created_date) AS created_weekday,
        HOUR(created_date) AS created_hour,
        closure_duration_minutes,
        NTILE(4) OVER (ORDER BY closure_duration_minutes) AS duration_quartile
    FROM fact_work_order
    WHERE closure_duration_minutes IS NOT NULL
)
-- 比较最短25%和最长25%的工单构成。
SELECT
    CASE
        WHEN duration_quartile = 1 THEN 'Short'
        WHEN duration_quartile = 4 THEN 'Long'
    END AS duration_group,
    complaint_type,
    agency,
    borough,
    open_data_channel_type,
    created_weekday,
    created_hour,
    COUNT(*) AS work_order_count
FROM duration_orders
WHERE duration_quartile IN (1, 4)
GROUP BY
    duration_group,
    complaint_type,
    agency,
    borough,
    open_data_channel_type,
    created_weekday,
    created_hour
ORDER BY work_order_count DESC;

Running query in 'mysql+pymysql://root:***@localhost:3306/nyc_311_project?charset=utf8mb4'

42339 rows affected.

duration_group,complaint_type,agency,borough,open_data_channel_type,created_weekday,created_hour,work_order_count
Short,Noise - Residential,NYPD,BRONX,MOBILE,2,0,383
Short,Noise - Residential,NYPD,BRONX,MOBILE,2,8,339
Short,Noise - Residential,NYPD,BRONX,MOBILE,1,13,302
Short,Noise - Residential,NYPD,BRONX,MOBILE,1,0,275
Short,Noise - Residential,NYPD,BRONX,MOBILE,7,20,266
Short,Noise - Residential,NYPD,BRONX,MOBILE,7,21,219
Short,Noise - Residential,NYPD,BRONX,MOBILE,2,9,218
Short,Noise - Residential,NYPD,BRONX,MOBILE,1,12,213
Short,Noise - Residential,NYPD,BRONX,MOBILE,1,14,212
Short,Noise - Residential,NYPD,BRONX,MOBILE,1,11,204


前50个组合中，Short合计6343条，Long合计189条。数量最大的组合主要是NYPD在Bronx通过MOBILE提交的住宅噪声，并分散在不同星期和小时。

由于按组合数量统一排序，前50行主要展示Short，不能据此概括全部Long工单。星期和小时只表示提交时间，不代表机构工作时间。

7. 提交渠道与问题类型、提交时间和关闭时间是否有关？

In [9]:
%%sql
-- 按渠道、工单类型、机构和提交时间组合统计。
SELECT
    open_data_channel_type,
    complaint_type,
    agency,
    DAYOFWEEK(created_date) AS created_weekday,
    HOUR(created_date) AS created_hour,
    COUNT(*) AS work_order_count,
    COUNT(closed_date) AS closed_order_count,
    ROUND(COUNT(closed_date) / COUNT(*) * 100, 2) AS closed_percentage,
    ROUND(AVG(closure_duration_minutes) / 60, 2) AS average_closure_hours
FROM fact_work_order
WHERE complaint_type IS NOT NULL
GROUP BY
    open_data_channel_type,
    complaint_type,
    agency,
    created_weekday,
    created_hour
ORDER BY work_order_count DESC;

Running query in 'mysql+pymysql://root:***@localhost:3306/nyc_311_project?charset=utf8mb4'

25660 rows affected.

open_data_channel_type,complaint_type,agency,created_weekday,created_hour,work_order_count,closed_order_count,closed_percentage,average_closure_hours
ONLINE,Noise - Residential,NYPD,7,22,657,657,100.00,3.51
ONLINE,Noise - Residential,NYPD,7,23,613,613,100.00,3.67
ONLINE,Noise - Residential,NYPD,1,0,592,592,100.00,2.22
MOBILE,Noise - Residential,NYPD,1,0,553,553,100.00,1.5
MOBILE,Noise - Residential,NYPD,1,17,540,540,100.00,2.82
MOBILE,Noise - Residential,NYPD,1,16,510,510,100.00,3.19
MOBILE,Noise - Residential,NYPD,1,22,505,505,100.00,1.31
MOBILE,Noise - Residential,NYPD,7,23,481,481,100.00,2.26
MOBILE,Noise - Residential,NYPD,1,18,469,469,100.00,2.37
MOBILE,Noise - Residential,NYPD,2,0,468,468,100.00,0.73


前50行全部属于NYPD，共18493条；其中住宅噪声17181条，远多于街道噪声和违停。提交渠道只出现MOBILE和ONLINE，MOBILE合计14008条，ONLINE合计4485条。

这些是数量最大的50个时间组合，主要反映NYPD噪声工单，不能据此判断所有渠道与关闭时间的总体关系。后续看板需要固定机构和工单类型后再比较渠道。